File: djinn/standardize.py
import re
import rich_click as click
from djinn.utils.standardize_bam import std_bam
from djinn.utils.standardize_fastq import std_fastq
from djinn.utils.file_ops import validate_fq_sam
@click.command(no_args_is_help = True, context_settings={"allow_interspersed_args" : False}, epilog = "Documentation: https://pdimens.github.io/djinn/standardize/#fastq")
@click.option('-s', '--style', type = click.Choice(["haplotagging", "stlfr", "tellseq", "10x"], case_sensitive=False), help = 'Change the barcode style')
@click.argument('prefix', metavar="output_prefix", type = str, required = True, nargs=1)
@click.argument('inputs', type = click.Path(exists=True,dir_okay=False,readable=True,resolve_path=True), required = True, callback = validate_fq_sam, nargs=-1)
def standardize(prefix, inputs, style):
    """
    Move barcodes to `BX`/`VX` sequence header tags

    This conversion moves the barcode to the `BX:Z` tag in fastq/bam records, maintaining the same barcode type by default (auto-detected).
    See the documentation for a deeper look into the location and format expectations for different linked-read technologies.
    Also writes a `VX:i` tag to describe barcode validation `0` (invalid) or `1` (valid). Use `djinn fastq`
    if your fastq data is in 10X format, as this command will not work on 10X format (i.e. barcode is the first 16 bases of read 1).
    Use `--style` to also convert the barcode to a different style (`haplotagging`, `stlfr`, `tellseq`, `10X`).

    | Option         | Style                                        |
    |:---------------|:---------------------------------------------|
    | `haplotagging` | AxxCxxBxxDxx                                 |
    | `stlfr`        | 1_2_3                                        |
    | `tellseq`      | 18-base nucleotide (e.g. AGCCATGTACGTATGGTA) |
    | `10X`          | 16-base nucleotide (e.g. GGCTGAACACGTGCAG)   |
    """
    if len(inputs) == 1:
        std_bam(prefix, inputs[0], style)
    else:
        std_fastq(prefix, inputs[0], inputs[0], style)

File: djinn/cmd/standardize.go
// Copyright © 2026 Pavel Dimens | Github: pdimens
package cmd

import (
	"djinn/standardize"
	"fmt"
	"runtime"

	"github.com/spf13/cobra"
)

// preCmd represents the preprocess command
var stdXamCmd = &cobra.Command{
	Use:     "standardize [options] file.bam",
	Short:   "Convert standard linked-read format with BX:Z and VX:i tags",
	Example: "standardize -t 10 curratum.bam > curratum.std.bam",
	Long: "This conversion moves the barcode to the `BX:Z` tag in sam/bam records, maintaining the same barcode type. " +
		"See the documentation for a deeper look into the location and format expectations for different linked-read technologies. " +
		"Also writes a `VX:i` tag to describe barcode validation `0` (invalid) or `1` (valid). Writes to stdout.",
	DisableFlagsInUseLine: true,
	SilenceUsage:          true,
	Args: func(cmd *cobra.Command, args []string) error {
		if len(args) == 0 {
			fmt.Printf("%s", cmd.UsageString())
			return fmt.Errorf("please provide an input file")
		}
		//TODO not exact args, needs min/max
		if err := cobra.ExactArgs(1)(cmd, args); err != nil {
			return err
		}
		if err := filecheck(args[0]); err != nil {
			return err
		}
		if len(args) == 2 {
			if err := filecheck(args[1]); err != nil {
				return err
			}
		}
		return nil
	},
	RunE: func(cmd *cobra.Command, args []string) error {
		threads, err := cmd.Flags().GetInt("threads")
		if err != nil {
			return err
		}
		maxCores := runtime.NumCPU()
		// clamp between 1 and max system threads
		threads = min(maxCores, max(threads, 1))
		runtime.GOMAXPROCS(threads)

		sam, err := cmd.Flags().GetBool("sam")
		if err != nil {
			return err
		}
		if err != nil {
			return err
		}
		return standardize.Standardize(args[0], threads, sam)
	},
}

func init() {
	samCmd.AddCommand(stdXamCmd)

	//---Command line arguments-------------
	stdXamCmd.Flags().BoolP("sam", "S", false, "Output as SAM instead of BAM")
	stdXamCmd.Flags().IntP("threads", "@", 2, "Worker threads to use")
}
